src/microbe_stage/editor/EditorActionHistory.cs

```diff
@@ -138,8 +138,16 @@ public override bool Undo()
 
     public override void AddAction(TAction action)
     {
-        // TODO: check if the action can be combined (for example behaviour or rigidity slider subsequent edits should
+        // Check if the action can be combined (for example behaviour or rigidity slider subsequent edits should
         // combine) in a single step for undo
+        if (ActionIndex > 0 && action.Data.FirstOrDefault()!.WantsMerge)
```

Comment:
```suggestion
        if (ActionIndex > 0 && action.Data.First().WantsMergeWith(action.Data)
```

Also I'm slightly suspicious if you got the ActionIndex correct. Is it the case that we have the action index pointing *after* the latest performed action? Just to remind me. If that's the case then the logic is right here.

Reply:
It's correct.